File: bsp-testkit2/bsp-testkit/client/MockClient.scala
package org.jetbrains.bsp.testkit.client

import ch.epfl.scala.bsp4j._

import scala.collection.mutable

class MockClient extends BuildClient {
  private val showMessage = new mutable.ListBuffer[ShowMessageParams]
  private val logMessage = new mutable.ListBuffer[LogMessageParams]
  private val taskStart = new mutable.ListBuffer[TaskStartParams]
  private val taskProgress = new mutable.ListBuffer[TaskProgressParams]
  private val taskFinish = new mutable.ListBuffer[TaskFinishParams]
  private val publishDiagnostics = new mutable.ListBuffer[PublishDiagnosticsParams]
  private val didChangeBuildTarget = new mutable.ListBuffer[DidChangeBuildTarget]

  def getShowMessageNotifications: List[ShowMessageParams] = showMessage.toList

  def getLogMessageNotifications: List[LogMessageParams] = logMessage.toList

  def getTaskStartNotifications: List[TaskStartParams] = taskStart.toList

  def getTaskProgressNotifications: List[TaskProgressParams] = taskProgress.toList

  def getTaskFinishNotifications: List[TaskFinishParams] = taskFinish.toList

  def getPublishDiagnosticsNotifications: List[PublishDiagnosticsParams] = publishDiagnostics.toList

  def getDidChangeBuildTargetNotifications: List[DidChangeBuildTarget] = didChangeBuildTarget.toList

  override def onBuildShowMessage(params: ShowMessageParams): Unit = showMessage += params

  override def onBuildLogMessage(params: LogMessageParams): Unit = logMessage += params

  override def onBuildTaskStart(params: TaskStartParams): Unit = taskStart += params

  override def onBuildTaskProgress(params: TaskProgressParams): Unit = taskProgress += params

  override def onBuildTaskFinish(params: TaskFinishParams): Unit = taskFinish += params

  override def onBuildPublishDiagnostics(params: PublishDiagnosticsParams): Unit =
    publishDiagnostics += params

  override def onBuildTargetDidChange(params: DidChangeBuildTarget): Unit =
    didChangeBuildTarget += params

  override def onRunPrintStderr(params: PrintParams): Unit = {}

  override def onRunPrintStdout(params: PrintParams): Unit = {}
}

File: bsp-testkit2/bsp-testkit/client/MockClient.kt
package org.jetbrains.bsp.testkit.client

import org.jetbrains.bsp.protocol.DidChangeBuildTarget
import org.jetbrains.bsp.protocol.JoinedBuildClient
import org.jetbrains.bsp.protocol.LogMessageParams
import org.jetbrains.bsp.protocol.PrintParams
import org.jetbrains.bsp.protocol.PublishDiagnosticsParams
import org.jetbrains.bsp.protocol.PublishOutputParams
import org.jetbrains.bsp.protocol.ShowMessageParams
import org.jetbrains.bsp.protocol.TaskFinishParams
import org.jetbrains.bsp.protocol.TaskProgressParams
import org.jetbrains.bsp.protocol.TaskStartParams

open class MockClient : JoinedBuildClient {
  private val showMessage = ArrayList<ShowMessageParams>()
  private val logMessage = ArrayList<LogMessageParams>()
  private val taskStart = ArrayList<TaskStartParams>()
  private val taskProgress = ArrayList<TaskProgressParams>()
  private val taskFinish = ArrayList<TaskFinishParams>()
  private val publishDiagnostics = ArrayList<PublishDiagnosticsParams>()
  private val didChangeBuildTarget = ArrayList<DidChangeBuildTarget>()
  private val printStdout = ArrayList<PrintParams>()
  private val printStderr = ArrayList<PrintParams>()

  val showMessageNotifications: List<ShowMessageParams>
    get() = showMessage

  val logMessageNotifications: List<LogMessageParams>
    get() = logMessage

  val taskStartNotifications: List<TaskStartParams>
    get() = taskStart

  val taskProgressNotifications: List<TaskProgressParams>
    get() = taskProgress

  val taskFinishNotifications: List<TaskFinishParams>
    get() = taskFinish

  val publishDiagnosticsNotifications: List<PublishDiagnosticsParams>
    get() = publishDiagnostics

  val didChangeBuildTargetNotifications: List<DidChangeBuildTarget>
    get() = didChangeBuildTarget

  fun clearDiagnostics() {
    publishDiagnostics.clear()
  }

  override fun onBuildShowMessage(params: ShowMessageParams) {
    showMessage.add(params)
  }

  override fun onBuildLogMessage(params: LogMessageParams) {
    logMessage.add(params)
  }

  override fun onBuildTaskStart(params: TaskStartParams) {
    taskStart.add(params)
  }

  override fun onBuildTaskProgress(params: TaskProgressParams) {
    taskProgress.add(params)
  }

  override fun onBuildTaskFinish(params: TaskFinishParams) {
    taskFinish.add(params)
  }

  override fun onRunPrintStdout(params: PrintParams) {
    printStdout.add(params)
  }

  override fun onRunPrintStderr(params: PrintParams) {
    printStderr.add(params)
  }

  override fun onBuildPublishDiagnostics(params: PublishDiagnosticsParams) {
    publishDiagnostics.add(params)
  }

  override fun onBuildTargetDidChange(params: DidChangeBuildTarget) {
    didChangeBuildTarget.add(params)
  }

  override fun onBuildPublishOutput(params: PublishOutputParams) {}
}
